presto-main/src/main/java/com/facebook/presto/operator/LookupJoinPageBuilder.java

```diff
@@ -0,0 +1,141 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.operator;
+
+import com.facebook.presto.spi.Page;
+import com.facebook.presto.spi.PageBuilder;
+import com.facebook.presto.spi.block.Block;
+import com.facebook.presto.spi.block.DictionaryBlock;
+import com.facebook.presto.spi.block.LazyBlock;
+import com.facebook.presto.spi.type.Type;
+import it.unimi.dsi.fastutil.ints.IntArrayList;
+
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+import static com.facebook.presto.spi.block.PageBuilderStatus.DEFAULT_MAX_PAGE_SIZE_IN_BYTES;
+import static com.google.common.base.MoreObjects.toStringHelper;
+
+/**
+ * The page builder creates a page with dictionary blocks only
+ * the build side blocks are copied from the build pages;
+ * while the probe side blocks are dictionary views on top of the probe pages
+ * TODO: may extend the build side to dictionary as well
+ */
+public class LookupJoinPageBuilder
+{
+    private final IntArrayList probeIndexBuilder = new IntArrayList();
+    private final Set<Integer> probeIndices = new HashSet<>();
+    private final PageBuilder buildPageBuilder;
+    private final int buildOutputChannelCount;
+    private int probeBlockBytes;
+
+    public LookupJoinPageBuilder(List<Type> buildTypes)
+    {
+        this.buildPageBuilder = new PageBuilder(buildTypes);
+        this.buildOutputChannelCount = buildTypes.size();
+    }
+
+    public boolean isFull()
+    {
+        return probeBlockBytes + buildPageBuilder.getSizeInBytes() >= DEFAULT_MAX_PAGE_SIZE_IN_BYTES || buildPageBuilder.isFull();
+    }
+
+    public boolean isEmpty()
+    {
+        return probeIndexBuilder.isEmpty() && buildPageBuilder.isEmpty();
+    }
+
+    public void reset()
+    {
+        probeIndexBuilder.clear();
```

Comment:
There are other places in our code base doing something similar; e.g., `SliceDirectStreamReader#isNullVector`; will add a comment here in case it cause trouble.